extension/src/state_aggregate.rs

```diff
@@ -0,0 +1,342 @@
+//! SELECT duration_in('STOPPED', states) as run_time, duration_in('ERROR', states) as error_time FROM (
+//!   SELECT state_agg(time, state) as states FROM ...
+//! );
+//!
+//! Currently requires loading all data into memory in order to sort it by time.
+
+#![allow(non_camel_case_types)]
+
+use pgx::*;
+
+use aggregate_builder::aggregate;
+use flat_serialize::*;
+use state_aggregate::DurationInState;
+
+use crate::{
+    flatten,
+    palloc::{Inner, Internal},
+    pg_type,
+    raw::{bytea, TimestampTz},
+    ron_inout_funcs,
+};
+
+#[aggregate]
+impl state_agg {
+    type State = state_aggregate::TransState;
+
+    const PARALLEL_SAFE: bool = true;
+
+    fn transition(
+        state: Option<State>,
+        #[sql_type("timestamptz")] ts: TimestampTz,
+        #[sql_type("text")] value: Option<String>,
+    ) -> Option<State> {
+        let value = match value {
+            None => return state,
+            Some(value) => value,
+        };
+        let mut state = match state {
+            None => state_aggregate::TransState::new().into(),
+            Some(state) => state,
+        };
+        state.record(value, ts.into());
+        Some(state)
+    }
+
+    fn combine(a: Option<&State>, b: Option<&State>) -> Option<State> {
+        todo!("how can we force combine from a test?");
+        match (a, b) {
+            (None, None) => None,
+            (None, Some(only)) | (Some(only), None) => Some(only.clone()),
+            (Some(a), Some(b)) => {
+                let (mut a, mut b) = (a.clone(), b.clone());
+                a.append(&mut b);
+                Some(a)
+            }
+        }
+    }
+
+    fn serialize(state: &mut State) -> bytea {
+        crate::do_serialize!(state)
+    }
+
+    fn deserialize(bytes: bytea) -> State {
+        crate::do_deserialize!(bytes, state_aggregate::TransState)
+    }
+
+    fn finally(state: Option<&mut State>) -> Option<StateAgg<'static>> {
+        let durations = state
+            .map(|s| {
+                s.to_duration_map()
+                    .iter()
+                    .map(|(state, duration)| {
+                        DurationInState {
+                            duration: (*duration).into(),
+                            // TODO testing hack.  need to figure out how to get the state String into StateAgg
+                            state: hack(&state),
+                        }
+                    })
+                    .collect()
+            })
+            .unwrap_or_else(Vec::new);
+        Some(unsafe {
+            flatten!(StateAgg {
+                len: durations.len() as u64,
+                durations: (&*durations).into(),
+            })
+        })
+    }
+}
+
+#[pg_extern(immutable, parallel_safe)]
+pub fn duration_in(state: String, aggregate: Option<StateAgg>) -> i64 {
+    aggregate
+        .map(|aggregate| aggregate.get(&state))
+        .flatten()
+        .unwrap_or(0)
+}
+
+pg_type! {
+    #[derive(Debug)]
+    struct StateAgg<'input> {
```

Comment:
Ultimately you'll likely need to also store the start and end records for this range in order to implement `rollup` (combining multiple completed StateAggs).